workbench-core/datasets/src/dataSetService.ts

```diff
@@ -252,32 +258,54 @@ export class DataSetService {
    *
    * @returns an array of DataSet objects.
    */
-  public async listDataSets(authenticatedUser: AuthenticatedUser): Promise<DataSet[]> {
+  public async listDataSets(
+    authenticatedUser: AuthenticatedUser,
+    pageSize: number,
+    paginationToken: string | undefined
+  ): Promise<PaginatedResponse<DataSet>> {
     const metadata: Metadata = {
       actor: authenticatedUser,
       action: this.listDataSets.name,
       source: {
         serviceName: DataSetService.name
       }
     };
-
     try {
-      const response: DataSet[] = [];
-      const allDatasets = await this._dbProvider.listDataSets();
-
-      await Promise.all(
-        allDatasets.map(async (dataset) => {
-          try {
-            // throws a ForbiddenError if the user doesnt have permission on the dataset
-            await this._authzPlugin.isAuthorizedOnDataSet(dataset.id, 'read-only', authenticatedUser);
-            response.push(dataset);
-          } catch (error) {
-            if (!isForbiddenError(error)) {
-              throw error;
+      const authedDataSets: DataSet[] = [];
+
+      let lastPaginationToken = paginationToken;
+      do {
+        const pluginResponse = await this._dbProvider.listDataSets(pageSize, lastPaginationToken);
+
+        await Promise.all(
+          pluginResponse.data.map(async (dataset) => {
+            try {
+              await this._authzPlugin.isAuthorizedOnDataSet(dataset.id, 'read-only', authenticatedUser);
+              authedDataSets.push(dataset);
+            } catch (error) {
+              if (!isForbiddenError(error)) {
+                throw error;
+              }
             }
-          }
-        })
-      );
+          })
+        );
+        lastPaginationToken = pluginResponse.paginationToken;
+      } while (authedDataSets.length < pageSize && lastPaginationToken);
+
+      let sortedDataSets = authedDataSets.sort((a: DataSet, b: DataSet) => {
+        return Date.parse(b.createdAt) - Date.parse(a.createdAt);
+      });
+
+      if (authedDataSets.length > pageSize) {
+        sortedDataSets = sortedDataSets.slice(0, pageSize);
+        const lastKey = buildDynamoDBPkSk(sortedDataSets[pageSize - 1].id, resourceTypeToKey.dataset);
```

Comment:
the `resourceTypeKey.*` convention isn't valid in the `DataSets` context. The user is allowed to configure that when the `ddbDataSetMetadataPlugin` is discovered. You need to expose what the user passed to that plugin when it is constructed, or make it part of the `DataSetService`. In my implementation of this, I simply added a quick method to return the private property from `ddbDataSetMetadataPlugin` and exposed that method on the interface as well so other versions of the interface would provide a valid method to call.